verilog/tools/ls/symbol-table-handler.cc

```diff
@@ -0,0 +1,300 @@
+  buildstatus.insert(buildstatus.end(), resolvestatus.begin(),
+                     resolvestatus.end());
+  files_dirty_ = false;
+  return buildstatus;
+}
+
+bool SymbolTableHandler::LoadProjectFileList(absl::string_view current_dir) {
+  LOG(INFO) << __FUNCTION__;
+  if (!curr_project_) return false;
+  if (filelist_path_.empty()) {
+    // search for FileList file up the directory hierarchy
+    std::string projectpath = FindFileList(current_dir);
+    if (projectpath.empty()) {
+      filelist_path_ = "";
+      last_filelist_update_ = {};
+      return false;
+    }
+    LOG(INFO) << "Found file list under " << projectpath;
+    filelist_path_ = projectpath;
+  }
+  if (!last_filelist_update_) {
+    last_filelist_update_ = std::filesystem::last_write_time(filelist_path_);
+  } else if (*last_filelist_update_ ==
+             std::filesystem::last_write_time(filelist_path_)) {
+    // filelist file is unchanged, keeping it
+    return true;
+  }
+  LOG(INFO) << "Updating the filelist";
+  // fill the FileList object
+  FileList filelist;
+  if (absl::Status status = AppendFileListFromFile(filelist_path_, &filelist);
+      !status.ok()) {
+    // if failed to parse
+    LOG(WARNING) << "Failed to parse file list in " << filelist_path_ << ":  "
+                 << status;
+    filelist_path_ = "";
+    last_filelist_update_ = {};
+    return false;
+  }
+  // add directory containing filelist to includes
+  // TODO (glatosinski): should we do this?
+  const absl::string_view filelist_dir = verible::file::Dirname(filelist_path_);
+  curr_project_->AddIncludePath(filelist_dir);
+  LOG(INFO) << "Adding \"" << filelist_dir << "\" to include directories";
+  // update include directories in project
+  for (const auto &incdir : filelist.preprocessing.include_dirs) {
+    LOG(INFO) << "Adding include path:  " << incdir;
+    curr_project_->AddIncludePath(incdir);
+  }
+  // add files from file list to the project
+  for (const auto &file_in_project : filelist.file_paths) {
+    const std::string canonicalized =
+        std::filesystem::path(file_in_project).lexically_normal().string();
+    auto source = curr_project_->OpenTranslationUnit(canonicalized);
+    if (!source.ok()) source = curr_project_->OpenIncludedFile(canonicalized);
+    if (!source.ok()) {
+      LOG(WARNING) << "File included in " << filelist_path_
+                   << " not found:  " << canonicalized << ":  "
+                   << source.status();
+      continue;
+    }
+    LOG(INFO) << "Creating symbol table for:  " << canonicalized;
+    BuildSymbolTableFor(*source.value());
+  }
+  return true;
+}
+
+bool IsStringViewContained(absl::string_view origin, absl::string_view substr) {
+  const int from = std::distance(origin.begin(), substr.begin());
+  const int to = std::distance(origin.begin(), substr.end());
+  if (from < 0) return false;
+  if (to > static_cast<int>(origin.length())) return false;
+  return true;
+}
+
+const SymbolTableNode *ScanReferenceComponents(
```

Comment:
For future consideration: if scanning takes a while, maybe we need to build an index in the symbol table.